File: rt/entropy_core.v2.c
/**
 * FreeLang v2.40 - Native Entropy Core (entropy_core.v2.c)
 *
 * Kernel-level entropy source for UUID v4 generation
 * - /dev/urandom 직접 호출
 * - RDRAND 하드웨어 난수생성기 (Intel/AMD)
 * - 128-bit 바이너리 직접 할당 (문자열 변환 없음)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>

/* CPU RDRAND 가용성 확인 */
#ifdef __x86_64__
  #define HAS_RDRAND 1
  #include <immintrin.h>
#else
  #define HAS_RDRAND 0
#endif

/**
 * 엔트로피 상태 저장소 (글로벌)
 * 런타임 초기화 시 한 번만 열기
 */
static int entropy_fd = -1;
static int entropy_source = 0;  // 0: /dev/urandom, 1: RDRAND, -1: not initialized

#define ENTROPY_SOURCE_URANDOM 0
#define ENTROPY_SOURCE_RDRAND 1
#define ENTROPY_SOURCE_HYBRID 2

/**
 * entropy_init()
 * /dev/urandom 또는 RDRAND 초기화
 * Return: 0 (success), -1 (failure)
 */
int entropy_init() {
  if (entropy_source != 0 && entropy_fd != -1) {
    return 0;  // Already initialized
  }

#ifdef HAS_RDRAND
  // CPU에 RDRAND 지원 확인 (cpuid 사용)
  unsigned int eax, edx;
  __cpuid(1, eax, eax, eax, edx);
  if (edx & (1 << 30)) {
    entropy_source = ENTROPY_SOURCE_RDRAND;
    return 0;
  }
#endif

  // Fallback: /dev/urandom
  entropy_fd = open("/dev/urandom", O_RDONLY);
  if (entropy_fd < 0) {
    perror("Failed to open /dev/urandom");
    return -1;
  }

  entropy_source = ENTROPY_SOURCE_URANDOM;
  return 0;
}

/**
 * entropy_generate_128bit(buf[16])
 * 128-bit (16 바이트) 난수 직접 할당
 *
 * UUID v4 형식 (RFC 4122):
 * - Bits 0-3: version (0100 = v4)
 * - Bits 4-7: variant (10xx = RFC 4122)
 * - Bits 8-127: random
 *
 * Return: 0 (success), -1 (failure)
 */
int entropy_generate_128bit(unsigned char* buf, int buflen) {
  if (buflen < 16) {
    return -1;
  }

  if (entropy_source < 0) {
    if (entropy_init() < 0) {
      return -1;
    }
  }

#ifdef HAS_RDRAND
  if (entropy_source == ENTROPY_SOURCE_RDRAND) {
    // Intel RDRAND 사용
    unsigned long long *ptr = (unsigned long long *)buf;

    // 64-bit × 2 = 128-bit
    if (!_rdrand64_step(&ptr[0])) {
      goto fallback_urandom;
    }
    if (!_rdrand64_step(&ptr[1])) {
      goto fallback_urandom;
    }

    // UUID v4 바이트 조정 (RFC 4122)
    buf[6] = (buf[6] & 0x0F) | 0x40;  // version 4
    buf[8] = (buf[8] & 0x3F) | 0x80;  // variant

    return 0;
  }

fallback_urandom:
#endif

  // /dev/urandom에서 16바이트 읽기
  ssize_t n = read(entropy_fd, buf, 16);
  if (n < 16) {
    perror("Failed to read from /dev/urandom");
    return -1;
  }

  // UUID v4 바이트 조정
  buf[6] = (buf[6] & 0x0F) | 0x40;  // version 4
  buf[8] = (buf[8] & 0x3F) | 0x80;  // variant

  return 0;
}

/**
 * entropy_cleanup()
 * 엔트로피 리소스 정리
 */
void entropy_cleanup() {
  if (entropy_fd >= 0) {
    close(entropy_fd);
    entropy_fd = -1;
  }
  entropy_source = -1;
}

/**
 * entropy_stats()
 * 현재 엔트로피 소스 상태 반환
 * Return: 0 (urandom), 1 (rdrand), -1 (not initialized)
 */
int entropy_stats() {
  if (entropy_source < 0) {
    entropy_init();
  }
  return entropy_source;
}

/* 테스트 바이너리 (선택사항) */
#ifdef ENTROPY_STANDALONE_TEST
#include <stdio.h>

int main() {
  unsigned char uuid[16];

  if (entropy_init() < 0) {
    fprintf(stderr, "Failed to initialize entropy\n");
    return 1;
  }

  if (entropy_generate_128bit(uuid, 16) < 0) {
    fprintf(stderr, "Failed to generate UUID\n");
    return 1;
  }

  printf("Generated UUID (hex): ");
  for (int i = 0; i < 16; i++) {
    printf("%02x", uuid[i]);
  }
  printf("\n");

  printf("Entropy source: %d\n", entropy_stats());

  entropy_cleanup();
  return 0;
}
#endif

File: rt/entropy_core.v2.ts
/**
 * FreeLang v2.40 - Entropy Core Binding (entropy_core.v2.ts)
 *
 * Native C 엔트로피 소스 바인딩 (koffi FFI)
 * - /dev/urandom 또는 RDRAND 사용
 * - 128-bit 순수 바이너리 직접 생성
 */

import * as fs from 'fs';
import * as path from 'path';

let entropyLib: any = null;
let entropyInitialized = false;

/**
 * 엔트로피 라이브러리 동적 로드
 * 컴파일된 entropy_core.so/dylib를 로드
 */
export function loadEntropyCore(): boolean {
  try {
    // 먼저 koffi (Node.js FFI 라이브러리) 확인
    let koffi: any;
    try {
      koffi = require('koffi');
    } catch (e) {
      console.warn('[entropy] koffi not available, using pure JS fallback');
      return true;  // koffi 없으면 JS fallback 사용
    }

    // 엔트로피 라이브러리 경로 찾기
    const libPaths = [
      path.join(__dirname, 'entropy_core.so'),
      path.join(__dirname, '..', 'build', 'entropy_core.so'),
      path.join(__dirname, '..', 'rt', 'entropy_core.so'),
    ];

    for (const libPath of libPaths) {
      if (fs.existsSync(libPath)) {
        console.log(`[entropy] Loading ${libPath}`);

        // C 함수 정의
        const lib = koffi.load(libPath);
        entropyLib = {
          entropy_init: koffi.stdcall('int entropy_init()', lib),
          entropy_generate_128bit: koffi.stdcall(
            'int entropy_generate_128bit(uchar* buf, int buflen)',
            lib
          ),
          entropy_cleanup: koffi.stdcall('void entropy_cleanup()', lib),
          entropy_stats: koffi.stdcall('int entropy_stats()', lib),
        };

        // 초기화
        if (entropyLib.entropy_init() === 0) {
          entropyInitialized = true;
          console.log('[entropy] Initialized successfully');
          return true;
        }
      }
    }

    console.warn('[entropy] entropy_core.so not found, using pure JS fallback');
    return false;
  } catch (error) {
    console.warn(`[entropy] FFI loading failed: ${error}`);
    return false;
  }
}

/**
 * Pure JavaScript Fallback (Node.js crypto)
 * 엔트로피 라이브러리를 로드할 수 없을 때 사용
 */
function generateUUIDv4Fallback(): Buffer {
  const crypto = require('crypto');
  const buf = crypto.randomBytes(16);

  // UUID v4 마킹
  buf[6] = (buf[6] & 0x0f) | 0x40;  // version 4
  buf[8] = (buf[8] & 0x3f) | 0x80;  // variant

  return buf;
}

/**
 * entropy_init_global()
 * 전역 엔트로피 시스템 초기화
 */
export function entropy_init_global(): boolean {
  if (entropyInitialized) return true;

  const loaded = loadEntropyCore();
  if (!loaded) {
    console.log('[entropy] Using JS fallback');
  }

  return true;
}

/**
 * entropy_generate_128bit()
 * 128-bit 순수 바이너리 생성 (C 또는 JS)
 * Return: Buffer (16 bytes)
 */
export function entropy_generate_128bit(): Buffer {
  if (!entropyInitialized) {
    entropy_init_global();
  }

  // Native C 라이브러리 사용 가능
  if (entropyLib) {
    try {
      const buf = Buffer.alloc(16);
      const result = entropyLib.entropy_generate_128bit(buf, 16);

      if (result === 0) {
        return buf;
      }
    } catch (error) {
      console.warn(`[entropy] Native generation failed: ${error}, fallback to JS`);
    }
  }

  // Fallback to JS
  return generateUUIDv4Fallback();
}

/**
 * entropy_format_hex(buf)
 * 16바이트 바이너리 → 16진수 문자열 (옵션)
 * uuid: 550e8400-e29b-41d4-a716-446655440000
 */
export function entropy_format_hex(buf: Buffer): string {
  if (buf.length !== 16) {
    throw new Error('Invalid buffer size');
  }

  return [
    buf.slice(0, 4).toString('hex'),
    buf.slice(4, 6).toString('hex'),
    buf.slice(6, 8).toString('hex'),
    buf.slice(8, 10).toString('hex'),
    buf.slice(10, 16).toString('hex'),
  ].join('-');
}

/**
 * entropy_stats()
 * 엔트로피 소스 상태 조회
 * Return: { source: 'urandom'|'rdrand'|'js_fallback', initialized: boolean }
 */
export function entropy_stats(): {
  source: string;
  initialized: boolean;
  lib: boolean;
} {
  let source = 'js_fallback';

  if (entropyLib && entropyInitialized) {
    const stats = entropyLib.entropy_stats();
    if (stats === 0) source = 'urandom';
    else if (stats === 1) source = 'rdrand';
  }

  return {
    source,
    initialized: entropyInitialized,
    lib: !!entropyLib,
  };
}

/**
 * entropy_cleanup()
 * 리소스 정리
 */
export function entropy_cleanup() {
  if (entropyLib && entropyLib.entropy_cleanup) {
    try {
      entropyLib.entropy_cleanup();
    } catch (error) {
      console.warn(`[entropy] Cleanup failed: ${error}`);
    }
  }
  entropyInitialized = false;
}

// 모듈 로드 시 자동 초기화
process.on('exit', () => {
  entropy_cleanup();
});
